File: 1818-maximum-score-from-removing-substrings/1818-maximum-score-from-removing-substrings.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        string top, bot;
        int top_score, bot_score;

        if(x > y) {
            top = "ab";
            top_score = x;
            bot = "ba";
            bot_score = y;
        }
        else {
            top = "ba";
            top_score = y;
            bot = "ab";
            bot_score = x;
        }

        vector<char> stack;
        int ans = 0;
        for(char c : s) {
            if(c == top[1] && !stack.empty() && stack.back() == top[0]) {
               ans += top_score;
               stack.pop_back();
            }
            else {
                stack.push_back(c);
            }
        }

        vector<char> stack1;
        for(char c : stack) {
            if(c == bot[1] && !stack1.empty() && stack1.back() == bot[0]) {
                ans += bot_score;
                stack1.pop_back();
            }
            else {
                stack1.push_back(c);
            }
        }

        return ans;
    }
};

File: 1818-maximum-score-from-removing-substrings/1818-maximum-score-from-removing-substrings.js
/**
 * @param {string} s
 * @param {number} x
 * @param {number} y
 * @return {number}
 */
var maximumGain = function(s, x, y) {
    let top, bot, topScore, botScore;

    if (x > y) {
        top = "ab";
        topScore = x;
        bot = "ba";
        botScore = y;
    } else {
        top = "ba";
        topScore = y;
        bot = "ab";
        botScore = x;
    }

    let stack = [];
    let ans = 0;
    for (let c of s) {
        if (c === top[1] && stack.length && stack[stack.length - 1] === top[0]) {
            ans += topScore;
            stack.pop();
        } else {
            stack.push(c);
        }
    }

    let stack1 = [];
    for (let c of stack) {
        if (c === bot[1] && stack1.length && stack1[stack1.length - 1] === bot[0]) {
            ans += botScore;
            stack1.pop();
        } else {
            stack1.push(c);
        }
    }

    return ans;
};